File: web/src/api/stats.js
import { request } from './client'

export function summary() {
  return request('/stats/summary')
}

export function byYear() {
  return request('/stats/by-year')
}

export function byMonth(year) {
  return request(`/stats/by-month${year ? `?year=${year}` : ''}`)
}

export function byLanguage(scope) {
  return request(`/stats/by-language${scope ? `?scope=${scope}` : ''}`)
}

export function topAuthors(scope, limit) {
  const params = new URLSearchParams()
  if (scope) params.set('scope', scope)
  if (limit) params.set('limit', String(limit))
  const qs = params.toString()
  return request(`/stats/top-authors${qs ? `?${qs}` : ''}`)
}

export function streak() {
  return request('/stats/streak')
}

File: internal/stats/stats.go
// Package stats computes each user's reading statistics — summary counts,
// yearly/monthly breakdowns, language and author rankings, and reading
// streaks — over their library.
package stats

// Summary is the headline numbers shown at the top of the Reports screen.
// CurrentYear is the server's UTC year, echoed back so the client never has
// to reconcile its own clock against the server's idea of "this year".
type Summary struct {
	TotalRead   int
	Reading     int
	Backlog     int
	ThisYear    int
	LastYear    int
	CurrentYear int
	UndatedRead int
}

// YearCount is one year's count of read books with a finished_at date.
// Year is a string, not an int: it is an opaque label, not a number to do
// arithmetic on, matching how the rest of the codebase treats it.
type YearCount struct {
	Year  string
	Count int
}

// MonthCount is one calendar month's count within a single year. Month is
// 1-12.
type MonthCount struct {
	Month int
	Count int
}

// LanguageCount is one language's count of read books within a scope.
type LanguageCount struct {
	Language string
	Count    int
}

// AuthorCount is one author's count of read books within a scope. A book
// with several authors counts once per author.
type AuthorCount struct {
	Author string
	Count  int
}
